lib/templates.php

```diff
@@ -231,3 +224,59 @@ function filter_rest_wp_template_query( $args, $request ) {
 	return $args;
 }
 add_filter( 'rest_wp_template_query', 'filter_rest_wp_template_query', 99, 2 );
+
+function gutenberg_create_template_auto_drafts() {
+	$create_auto_drafts = false;
+
+	$theme             = wp_get_theme();
+	$stylesheet        = $theme->get_stylesheet();
+	$parent_theme      = $theme->parent();
+	$parent_stylesheet = $parent_theme ? $parent_theme->get_stylesheet() : null;
+
+	$last_auto_drafts_versions = get_option( 'gutenberg_last_template_auto_drafts_theme_versions', array() );
+	if ( isset( $last_auto_drafts_versions[ $stylesheet ] ) ) {
+		$last_version = $last_auto_drafts_versions[ $stylesheet ];
+		if ( version_compare( $theme->Version, $last_version, '>' ) ) {
+			$create_auto_drafts = true;
+		}
+	} else {
+		$create_auto_drafts = true;
+	}
+
+	if ( $parent_theme ) {
+		if ( isset( $last_auto_drafts_versions[ $parent_stylesheet ] ) ) {
+			$last_version = $last_auto_drafts_versions[ $parent_stylesheet ];
+			if ( version_compare( $parent_theme->Version, $last_version, '>' ) ) {
+				$create_auto_drafts = true;
+			}
+		} else {
+			$create_auto_drafts = true;
+		}
+	}
+
+	if ( ! $create_auto_drafts ) {
+		return;
+	}
+
+	$block_template_files = gutenberg_get_template_paths();
+	foreach ( $block_template_files as $path ) {
+		$template_type = basename( $path, '.html' );
+		gutenberg_find_template_post_and_parts( $template_type, array( $template_type ) );
+	}
+
+	$last_auto_drafts_versions[ $stylesheet ] = $theme->Version;
+	if ( $parent_theme ) {
+		$last_auto_drafts_versions[ $parent_stylesheet ] = $parent_theme->Version;
+	}
+	update_option( 'gutenberg_last_template_auto_drafts_theme_versions', $last_auto_drafts_versions );
+}
+add_action( 'wp_loaded', 'gutenberg_create_template_auto_drafts' );
+
+function gutenberg_create_template_auto_drafts_on_shutdown() {
+	if ( ! gutenberg_is_block_editor() ) {
+		return;
+	}
+
+	gutenberg_create_template_auto_drafts();
+}
+register_shutdown_function( 'gutenberg_create_template_auto_drafts_on_shutdown' );
```

Comment:
Curious what the purpose of adding this on shutdown is?

Reply:
The idea is to run it once the response is sent to the user. So we don't delay TTFB.

When we open the site editor we re-create the auto-drafts in the background. If any of the template files changed then they will be updated in the database as well. (See `Reconciliation after updating files`)